compiler/optimizer/OMRSimplifierHandlers.cpp

```diff
@@ -12845,195 +12845,292 @@ TR::Node *su2dSimplifier(TR::Node * node, TR::Block * block, TR::Simplifier * s)
    }
 
 
+bool canTransformOp(bool canTransformAdd, bool canTransformSub, TR::Node *node, TR::Simplifier *s)
+   {
+   if (!(canTransformAdd || canTransformSub))
+      {
+      if (s->trace())
+         traceMsg(s->comp(),
+                  "\nEliminating add/sub under order compare node n%dn failed due to overflow\n",
+                  node->getGlobalIndex());
+
+      return false;
+      }
+   return true;
+   }
+
+// Overflow/Underflow handling:
+// Must detect overflow without causing overflow:
+// Equation: add operation: x + Const1 < Const2 --> x < Const2 - Const1
+//          sub operation: x - Const1 < Const2 --> x < Const2 + Const1
+
+//Signed Constants may be negative, therefore underflow and overflow are possible for both add and sub operations
+bool isNotSignedOverflow(bool isAddOp, bool isSubOp, uint64_t oldConst1, uint64_t oldConst2, uint64_t signedMax){
+   return (isAddOp && !(oldConst1 < 0 && (oldConst2 > signedMax + oldConst1)))
```

Comment:
how can oldConst2 be 0 in an unsigned representation?

Reply:
This is signed representation

Reply:
This comment links with https://github.com/eclipse/omr/pull/4305#discussion_r477664809
I realize this should be in signed representation, and in that case, I cannot simplify according to your suggestion in the linked comment.

Reply:
signed overflow is not defined in the C++ specification the math has to be done with unsigned math.

Reply:
Ah I see.In that case,  I checked to see if the constant>(unsigned)max to check if the Leftmost bit was set rather than playing with 0. I was then able to consolidate both functions into one check for overflow and one check for underflow.